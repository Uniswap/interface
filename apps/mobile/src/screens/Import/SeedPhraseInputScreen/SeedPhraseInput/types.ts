import { NativeSyntheticEvent, StyleProp, ViewStyle } from 'react-native'

import { TestIDType } from 'uniswap/src/test/fixtures/testIDs'

export enum StringKey {
  InputPlaceholder = 'inputPlaceholder',
  PasteButton = 'pasteButton',
  ErrorInvalidWord = 'errorInvalidWord',
  ErrorPhraseLength = 'errorPhraseLength',
  ErrorWrongPhrase = 'errorWrongPhrase',
  ErrorInvalidPhrase = 'errorInvalidPhrase',
}

type MnemonicStoredEvent = {
  mnemonicId: string
}

type InputValidatedEvent = {
  canSubmit: boolean
}

type HeightMeasuredEvent = {
  height: number
}

export type NativeSeedPhraseInputProps = {
  targetMnemonicId?: string
  testID?: TestIDType
  strings: Record<StringKey, string>
  onInputValidated: (e: NativeSyntheticEvent<InputValidatedEvent>) => void
  onMnemonicStored: (e: NativeSyntheticEvent<MnemonicStoredEvent>) => void
  // Only needed on iOS to determine when paste permission modal is open, which triggers inactive app state
  // And we need to prevent splash screen from appearing
  onPasteStart: () => void
  onPasteEnd: () => void
  onSubmitError: () => void
}

export type NativeSeedPhraseInputInternalProps = NativeSeedPhraseInputProps & {
  onHeightMeasured: (e: NativeSyntheticEvent<HeightMeasuredEvent>) => void
  style: StyleProp<ViewStyle>
}

export type NativeSeedPhraseInputRef = { handleSubmit: () => void; focus: () => void; blur: () => void }
