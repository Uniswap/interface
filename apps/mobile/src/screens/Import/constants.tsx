import React from 'react'
import { OSDynamicCloudIcon, PaperStack, Passkey } from 'ui/src/components/icons'
import { AppTFunction } from 'ui/src/i18n/types'
import { ElementName, ElementNameType } from 'uniswap/src/features/telemetry/constants'
import { TestID, TestIDType } from 'uniswap/src/test/fixtures/testIDs'
import { ImportType } from 'uniswap/src/types/onboarding'
import { OnboardingScreens } from 'uniswap/src/types/screens/mobile'
import { isAndroid } from 'utilities/src/platform'

export interface ImportMethodOption {
  title: (t: AppTFunction) => string
  blurb: (t: AppTFunction) => string
  icon: React.ReactNode
  nav: OnboardingScreens
  importType: ImportType
  name: ElementNameType
  testID: TestIDType
}

/**
 * Sign in with Passkey
 */
export const passKeySignInOption: ImportMethodOption = {
  title: (t: AppTFunction) => t('onboarding.import.selectMethod.passkey.title'),
  blurb: (t: AppTFunction) => t('onboarding.import.selectMethod.passkey.subtitle'),
  icon: <Passkey color="$accent1" size="$icon.18" />,
  nav: OnboardingScreens.PasskeyImport,
  importType: ImportType.Passkey,
  name: ElementName.OnboardingPasskey,
  testID: TestID.OnboardingPasskey,
}

/**
 * Import with Seed Phrase
 */
export const seedPhraseImportOption: ImportMethodOption = {
  title: (t: AppTFunction) => t('onboarding.import.selectMethod.recoveryPhrase.title'),
  blurb: (t: AppTFunction) => t('onboarding.import.selectMethod.recoveryPhrase.subtitle'),
  icon: <PaperStack color="$accent1" size="$icon.18" strokeWidth={1.5} />,
  nav: OnboardingScreens.SeedPhraseInput,
  importType: ImportType.SeedPhrase,
  name: ElementName.OnboardingImportSeedPhrase,
  testID: TestID.OnboardingImportSeedPhrase,
}

/**
 * Restore wallet with Seed Phrase.
 *
 * This will not allow a seed phrase that does not match the active wallet.
 */
export const restoreWalletWithSeedPhraseOption: ImportMethodOption = {
  title: (t: AppTFunction) => t('onboarding.import.method.restoreSeedPhrase.wallet.title'),
  blurb: (t: AppTFunction) => t('onboarding.import.method.restoreSeedPhrase.wallet.desc'),
  icon: <PaperStack color="$accent1" size="$icon.18" strokeWidth={1.5} />,
  nav: OnboardingScreens.SeedPhraseInput,
  importType: ImportType.RestoreMnemonic,
  name: ElementName.OnboardingImportSeedPhrase,
  testID: TestID.OnboardingImportSeedPhrase,
}

/**
 * Import from Cloud (iCloud/GDrive) Backup
 */
export const importFromCloudBackupOption: ImportMethodOption = {
  title: (t: AppTFunction) => t('onboarding.import.method.restore.title'),
  blurb: (t: AppTFunction) =>
    isAndroid
      ? t(`onboarding.import.method.restore.message.android`)
      : t(`onboarding.import.method.restore.message.ios`),
  icon: <OSDynamicCloudIcon color="$accent1" size="$icon.18" />,
  nav: OnboardingScreens.RestoreCloudBackup,
  importType: ImportType.Restore,
  name: ElementName.RestoreFromCloud,
  testID: TestID.RestoreFromCloud,
}

/**
 * Restore from Cloud (iCloud/GDrive) Backup. Checks if the cloud backup matches the active wallet.
 */
export const restoreFromCloudBackupOption: ImportMethodOption = {
  title: (t: AppTFunction) =>
    isAndroid
      ? t('onboarding.import.method.restoreSeedPhrase.cloud.title.android')
      : t('onboarding.import.method.restoreSeedPhrase.cloud.title.ios'),
  blurb: (t: AppTFunction) =>
    isAndroid
      ? t('onboarding.import.method.restoreSeedPhrase.cloud.desc.android')
      : t('onboarding.import.method.restoreSeedPhrase.cloud.desc.ios'),
  icon: <OSDynamicCloudIcon color="$accent1" size="$icon.18" />,
  nav: OnboardingScreens.RestoreCloudBackupLoading,
  importType: ImportType.RestoreMnemonic,
  name: ElementName.RestoreFromCloud,
  testID: TestID.RestoreFromCloud,
}
