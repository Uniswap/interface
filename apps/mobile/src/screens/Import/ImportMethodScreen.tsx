import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { Alert } from 'react-native'
import { useAppDispatch } from 'src/app/hooks'
import { OnboardingStackParamList } from 'src/app/navigation/types'
import Trace from 'src/components/Trace/Trace'
import { IS_ANDROID } from 'src/constants/globals'
import { isCloudStorageAvailable } from 'src/features/CloudBackup/RNCloudStorageBackupsManager'
import { OnboardingScreen } from 'src/features/onboarding/OnboardingScreen'
import { OptionCard } from 'src/features/onboarding/OptionCard'
import { ElementName } from 'src/features/telemetry/constants'
import { OnboardingScreens } from 'src/screens/Screens'
import { openSettings } from 'src/utils/linking'
import { useAddBackButton } from 'src/utils/useAddBackButton'
import { Flex, Icons, Text, TouchableArea, useSporeColors } from 'ui/src'
import EyeIcon from 'ui/src/assets/icons/eye.svg'
import { AppTFunction } from 'ui/src/i18n/types'
import { iconSizes } from 'ui/src/theme'
import { ImportType, OnboardingEntryPoint } from 'wallet/src/features/onboarding/types'
import {
  PendingAccountActions,
  pendingAccountActions,
} from 'wallet/src/features/wallet/create/pendingAccountsSaga'

interface ImportMethodOption {
  title: (t: AppTFunction) => string
  blurb: (t: AppTFunction) => string
  icon: React.ReactNode
  nav: OnboardingScreens
  importType: ImportType
  name: ElementName
}

const options: ImportMethodOption[] = [
  {
    title: (t: AppTFunction) => t('Import a wallet'),
    blurb: (t: AppTFunction) => t('Enter your recovery phrase from another crypto wallet'),
    icon: <Icons.PaperStack color="$accent1" size={18} strokeWidth={1.5} />,
    nav: OnboardingScreens.SeedPhraseInput,
    importType: ImportType.SeedPhrase,
    name: ElementName.OnboardingImportSeedPhrase,
  },
  {
    title: (t: AppTFunction) => t('Restore a wallet'),
    blurb: (t: AppTFunction) =>
      IS_ANDROID
        ? t(`Add wallets you’ve backed up to your Google Drive account`)
        : t(`Add wallets you’ve backed up to your iCloud account`),
    icon: <Icons.OSDynamicCloudIcon color="$accent1" size="$icon.24" />,
    nav: OnboardingScreens.RestoreCloudBackup,
    importType: ImportType.Restore,
    name: ElementName.RestoreFromCloud,
  },
]

type Props = NativeStackScreenProps<OnboardingStackParamList, OnboardingScreens.ImportMethod>

export function ImportMethodScreen({ navigation, route: { params } }: Props): JSX.Element {
  const { t } = useTranslation()
  const colors = useSporeColors()
  const dispatch = useAppDispatch()
  const entryPoint = params?.entryPoint

  useAddBackButton(navigation)

  const handleOnPressRestoreBackup = async (): Promise<void> => {
    const cloudStorageAvailable = await isCloudStorageAvailable()

    if (!cloudStorageAvailable) {
      Alert.alert(
        IS_ANDROID ? t('Google Drive not available') : t('iCloud Drive not available'),
        IS_ANDROID
          ? t(
              'Please verify that you are logged in to a Google account with Google Drive enabled on this device and try again.'
            )
          : t(
              'Please verify that you are logged in to an Apple ID with iCloud Drive enabled on this device and try again.'
            ),
        [
          { text: t('Go to settings'), onPress: openSettings, style: 'default' },
          { text: t('Not now'), style: 'cancel' },
        ]
      )
      return
    }

    navigation.navigate({
      name: OnboardingScreens.RestoreCloudBackupLoading,
      params: { importType: ImportType.Restore, entryPoint },
      merge: true,
    })
  }

  const handleOnPress = async (nav: OnboardingScreens, importType: ImportType): Promise<void> => {
    // Delete any pending accounts before entering flow.
    dispatch(pendingAccountActions.trigger(PendingAccountActions.Delete))

    if (importType === ImportType.Restore) {
      await handleOnPressRestoreBackup()
      return
    }

    navigation.navigate({
      name: nav,
      params: { importType, entryPoint },
      merge: true,
    })
  }

  const importOptions =
    entryPoint === OnboardingEntryPoint.Sidebar
      ? options.filter((option) => option.name !== ElementName.RestoreFromCloud)
      : options

  return (
    <OnboardingScreen title={t('How do you want to add your wallet?')}>
      <Flex grow gap="$spacing12" mt="$spacing4">
        {importOptions.map(({ title, blurb, icon, nav, importType, name }) => (
          <OptionCard
            key={'connection-option-' + title}
            hapticFeedback
            blurb={blurb(t)}
            elementName={name}
            icon={icon}
            title={title(t)}
            onPress={(): Promise<void> => handleOnPress(nav, importType)}
          />
        ))}
      </Flex>
      <Trace logPress element={ElementName.OnboardingImportBackup}>
        <TouchableArea alignItems="center" hitSlop={16} mb="$spacing12">
          <Flex row alignItems="center" gap="$spacing4">
            <EyeIcon
              color={colors.accent1.get()}
              height={iconSizes.icon20}
              strokeWidth="1.5"
              width={iconSizes.icon20}
            />
            <Text
              color="$accent1"
              variant="buttonLabel2"
              onPress={(): Promise<void> =>
                handleOnPress(OnboardingScreens.WatchWallet, ImportType.Watch)
              }>
              {t('Watch a wallet')}
            </Text>
          </Flex>
        </TouchableArea>
      </Trace>
    </OnboardingScreen>
  )
}
