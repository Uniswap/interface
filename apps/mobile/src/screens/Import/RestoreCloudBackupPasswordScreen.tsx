import { useFocusEffect } from '@react-navigation/core'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React, { useCallback, useRef, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Keyboard, TextInput } from 'react-native'
import { useAppDispatch, useAppSelector } from 'src/app/hooks'
import { OnboardingStackParamList } from 'src/app/navigation/types'
import { PasswordInput } from 'src/components/input/PasswordInput'
import { IS_ANDROID } from 'src/constants/globals'
import {
  incrementPasswordAttempts,
  resetLockoutEndTime,
  resetPasswordAttempts,
  setLockoutEndTime,
} from 'src/features/CloudBackup/passwordLockoutSlice'
import { restoreMnemonicFromCloudStorage } from 'src/features/CloudBackup/RNCloudStorageBackupsManager'
import { selectLockoutEndTime, selectPasswordAttempts } from 'src/features/CloudBackup/selectors'
import { OnboardingScreen } from 'src/features/onboarding/OnboardingScreen'
import { PasswordError } from 'src/features/onboarding/PasswordError'
import { ElementName } from 'src/features/telemetry/constants'
import { OnboardingScreens } from 'src/screens/Screens'
import { useAddBackButton } from 'src/utils/useAddBackButton'
import { Button, Flex, Text, TouchableArea } from 'ui/src'
import { ONE_HOUR_MS, ONE_MINUTE_MS } from 'utilities/src/time/time'
import { ImportType } from 'wallet/src/features/onboarding/types'
import { importAccountActions } from 'wallet/src/features/wallet/import/importAccountSaga'
import { ImportAccountType } from 'wallet/src/features/wallet/import/types'
import { NUMBER_OF_WALLETS_TO_IMPORT } from 'wallet/src/features/wallet/import/utils'

type Props = NativeStackScreenProps<
  OnboardingStackParamList,
  OnboardingScreens.RestoreCloudBackupPassword
>

/**
 * If the attempt count does not correspond to a lockout then returns undefined. Otherwise returns the lockout time based on attempts. The lockout time logic is as follows:
 * after 6 attempts, lock out for 5 minutes
 * after 10 attempts, lock out for 15 minutes
 * after 12 attempts and any subsequent multiple of 2, lock out for another 1hr
 */
function calculateLockoutEndTime(attemptCount: number): number | undefined {
  if (attemptCount < 6) {
    return undefined
  }
  if (attemptCount === 6) {
    return Date.now() + ONE_MINUTE_MS * 5
  }
  if (attemptCount < 10) {
    return undefined
  }
  if (attemptCount === 10) {
    return Date.now() + ONE_MINUTE_MS * 15
  }
  if (attemptCount < 12) {
    return undefined
  }
  if (attemptCount % 2 === 0) {
    return Date.now() + ONE_HOUR_MS
  }
  return undefined
}

function getLockoutTimeMessage(remainingLockoutTime: number): string {
  const minutes = Math.ceil(remainingLockoutTime / ONE_MINUTE_MS)
  if (minutes >= 60) {
    return '1 hour'
  }

  return minutes === 1 ? '1 minute' : `${minutes} minutes`
}

export function RestoreCloudBackupPasswordScreen({
  navigation,
  route: { params },
}: Props): JSX.Element {
  const { t } = useTranslation()
  const inputRef = useRef<TextInput>(null)
  const dispatch = useAppDispatch()

  const passwordAttemptCount = useAppSelector(selectPasswordAttempts)
  const lockoutEndTime = useAppSelector(selectLockoutEndTime)

  const isRestoringMnemonic = params.importType === ImportType.RestoreMnemonic

  const [enteredPassword, setEnteredPassword] = useState('')
  const [errorMessage, setErrorMessage] = useState<string | undefined>(undefined)

  const remainingLockoutTime = lockoutEndTime ? Math.max(0, lockoutEndTime - Date.now()) : 0
  const isLockedOut = remainingLockoutTime > 0

  useFocusEffect(
    useCallback(() => {
      if (isLockedOut) {
        setErrorMessage(
          t('Too many attempts. Try again in {{time}}.', {
            time: getLockoutTimeMessage(remainingLockoutTime),
          })
        )

        const timer = setTimeout(() => {
          setErrorMessage(undefined)
          dispatch(resetLockoutEndTime())
          inputRef.current?.focus()
        }, remainingLockoutTime)

        return () => clearTimeout(timer)
      }
    }, [isLockedOut, t, dispatch, remainingLockoutTime])
  )

  useAddBackButton(navigation)

  const onPasswordSubmit = async (): Promise<void> => {
    if (isLockedOut || enteredPassword.length === 0) return

    // Attempt to restore backup with encrypted mnemonic using password
    async function checkCorrectPassword(): Promise<void> {
      try {
        await restoreMnemonicFromCloudStorage(params.mnemonicId, enteredPassword)
        dispatch(
          importAccountActions.trigger({
            type: ImportAccountType.RestoreBackup,
            mnemonicId: params.mnemonicId,
            indexes: Array.from(Array(NUMBER_OF_WALLETS_TO_IMPORT).keys()),
          })
        )
        dispatch(resetPasswordAttempts())
        // restore flow is handled in saga after `restoreMnemonicComplete` is dispatched
        if (!isRestoringMnemonic) {
          navigation.navigate({ name: OnboardingScreens.SelectWallet, params, merge: true })
        }
      } catch (error) {
        dispatch(incrementPasswordAttempts())
        const updatedLockoutEndTime = calculateLockoutEndTime(passwordAttemptCount + 1)
        if (updatedLockoutEndTime) {
          dispatch(setLockoutEndTime({ lockoutEndTime: updatedLockoutEndTime }))
        } else {
          setErrorMessage(t('Invalid password. Please try again.'))
          inputRef.current?.focus()
        }
      }
    }

    await checkCorrectPassword()
    setEnteredPassword('')
    Keyboard.dismiss()
  }

  const navigateToEnterRecoveryPhrase = (): void => {
    navigation.replace(OnboardingScreens.SeedPhraseInput, params)
  }

  return (
    <OnboardingScreen
      subtitle={
        IS_ANDROID
          ? t('This password is required to recover your recovery phrase backup from Google Drive.')
          : t('This password is required to recover your recovery phrase backup from iCloud.')
      }
      title={
        IS_ANDROID
          ? t('Enter your Google Drive backup password')
          : t('Enter your iCloud backup password')
      }>
      <Flex>
        <PasswordInput
          ref={inputRef}
          autoFocus={!isLockedOut}
          editable={!isLockedOut}
          placeholder={t('Enter password')}
          value={enteredPassword}
          onChangeText={(newValue: string): void => {
            if (!isLockedOut) {
              setErrorMessage(undefined)
            }
            setEnteredPassword(newValue)
          }}
          onSubmitEditing={onPasswordSubmit}
        />
        {errorMessage && <PasswordError errorText={errorMessage} />}
      </Flex>
      <Flex>
        {isRestoringMnemonic && (
          <TouchableArea onPress={navigateToEnterRecoveryPhrase}>
            <Text color="$accent1" mb="$spacing12" textAlign="center" variant="buttonLabel3">
              {t('Enter your recovery phrase instead')}
            </Text>
          </TouchableArea>
        )}
        <Button
          disabled={!enteredPassword || isLockedOut}
          testID={ElementName.Submit}
          onPress={onPasswordSubmit}>
          {t('Continue')}
        </Button>
      </Flex>
    </OnboardingScreen>
  )
}
