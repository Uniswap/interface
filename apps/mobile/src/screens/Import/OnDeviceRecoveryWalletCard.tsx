import React, { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import { ViewProps } from 'react-native'
import { RecoveryWalletInfo, useOnDeviceRecoveryData } from 'src/screens/Import/useOnDeviceRecoveryData'
import { Button, Flex, FlexProps, Loader, Text, TouchableArea } from 'ui/src'
import { fonts, iconSizes } from 'ui/src/theme'
import { AccountIcon } from 'uniswap/src/features/accounts/AccountIcon'
import { useLocalizationContext } from 'uniswap/src/features/language/LocalizationContext'
import { NumberType } from 'utilities/src/format/types'
import { AddressDisplay } from 'wallet/src/components/accounts/AddressDisplay'

const cardProps: FlexProps & ViewProps = {
  borderRadius: '$rounded20',
  shadowColor: '$surface3',
  shadowOpacity: 0.04,
  shadowRadius: 10,
}

export function OnDeviceRecoveryWalletCard({
  mnemonicId,
  screenLoading,
  showAllWallets,
  onLoadComplete,
  onPressCard,
  onPressViewRecoveryPhrase,
}: {
  mnemonicId: string
  screenLoading: boolean
  showAllWallets: boolean
  onLoadComplete: (significantWalletCount: number) => void
  onPressCard: (walletInfos: RecoveryWalletInfo[]) => void
  onPressViewRecoveryPhrase: () => void
}): JSX.Element | null {
  const { t } = useTranslation()
  const { convertFiatAmountFormatted } = useLocalizationContext()

  const { recoveryWalletInfos, significantRecoveryWalletInfos, totalBalance, loading } =
    useOnDeviceRecoveryData(mnemonicId)

  const targetWalletInfos = showAllWallets ? recoveryWalletInfos.slice(0, 1) : significantRecoveryWalletInfos
  const firstWalletInfo = targetWalletInfos[0]
  const remainingWalletCount = targetWalletInfos.length - 1

  useEffect(() => {
    if (!loading && screenLoading) {
      onLoadComplete(significantRecoveryWalletInfos.length)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [loading, screenLoading])

  if (screenLoading || !firstWalletInfo) {
    return null
  }

  return (
    <TouchableArea onPress={() => onPressCard(targetWalletInfos)}>
      <Flex
        {...cardProps}
        centered
        backgroundColor="$surface1"
        borderColor="$surface2"
        borderWidth="$spacing1"
        gap="$spacing16"
        p="$spacing12"
      >
        <Flex centered row gap="$spacing12">
          <AccountIcon address={firstWalletInfo.address} size={iconSizes.icon36} />
          <Flex fill py={!remainingWalletCount ? fonts.body3.lineHeight / 2 : undefined}>
            <AddressDisplay
              address={firstWalletInfo.address}
              hideAddressInSubtitle={true}
              showAccountIcon={false}
              size={iconSizes.icon36}
              variant="subheading1"
            />
            {remainingWalletCount ? (
              <Text color="$neutral3" variant="body3">
                {t('onboarding.import.onDeviceRecovery.wallet.count', {
                  count: remainingWalletCount,
                })}
              </Text>
            ) : undefined}
          </Flex>

          <Text color="$neutral2" variant="body2">
            {convertFiatAmountFormatted(totalBalance, NumberType.PortfolioBalance)}
          </Text>
        </Flex>

        <Flex row>
          <Button emphasis="secondary" size="small" onPress={onPressViewRecoveryPhrase}>
            {t('onboarding.import.onDeviceRecovery.wallet.button')}
          </Button>
        </Flex>
      </Flex>
    </TouchableArea>
  )
}

const LOADING_MIN_OPACITY_SUBTRACT = 0.8

export function OnDeviceRecoveryWalletCardLoader({
  index,
  totalCount,
}: {
  index: number
  totalCount: number
}): JSX.Element {
  return (
    <Loader.Box height={120} opacity={1 - (LOADING_MIN_OPACITY_SUBTRACT * (index + 1)) / totalCount} {...cardProps} />
  )
}
