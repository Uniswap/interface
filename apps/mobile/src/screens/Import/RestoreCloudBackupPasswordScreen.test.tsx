import { RouteProp } from '@react-navigation/core'
import { NativeStackNavigationProp } from '@react-navigation/native-stack'
import React from 'react'
import { OnboardingStackParamList } from 'src/app/navigation/types'
import { RestoreCloudBackupPasswordScreen } from 'src/screens/Import/RestoreCloudBackupPasswordScreen'
import { OnboardingScreens } from 'src/screens/Screens'
import { render } from 'src/test/test-utils'
import { TamaguiProvider } from 'wallet/src/provider/tamagui-provider'

const setOptionsSpy = jest.fn()
const routeProp = { params: {} } as RouteProp<
  OnboardingStackParamList,
  OnboardingScreens.RestoreCloudBackupPassword
>

describe(RestoreCloudBackupPasswordScreen, () => {
  it('renders correctly', () => {
    const tree = render(
      <TamaguiProvider>
        <RestoreCloudBackupPasswordScreen
          navigation={
            {
              getState: () => ({
                index: 0,
              }),
              setOptions: setOptionsSpy,
            } as unknown as NativeStackNavigationProp<
              OnboardingStackParamList,
              OnboardingScreens.RestoreCloudBackupPassword,
              undefined
            >
          }
          route={routeProp}
        />
      </TamaguiProvider>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
