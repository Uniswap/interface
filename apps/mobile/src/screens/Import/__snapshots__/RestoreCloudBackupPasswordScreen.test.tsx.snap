// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestoreCloudBackupPasswordScreen renders correctly 1`] = `
<View
  sentry-label="Screen"
  style={
    {
      "alignItems": "stretch",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": "column",
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": 200,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          {
            "flex": 1,
            "justifyContent": "flex-end",
          },
          {
            "marginBottom": undefined,
          },
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      entering={BaseAnimationMock {}}
      exiting={BaseAnimationMock {}}
      style={
        {
          "alignItems": "stretch",
          "flexDirection": "column",
          "flexGrow": 1,
          "gap": 16,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "column",
            "gap": 12,
            "justifyContent": "center",
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          }
        }
      >
        <Text
          allowFontScaling={false}
          maxFontSizeMultiplier={1.2}
          style={
            {
              "color": "#222222",
              "fontFamily": "Basel-Book",
              "fontSize": 25,
              "lineHeight": 32,
              "paddingTop": 0,
              "textAlign": "center",
            }
          }
          suppressHighlighting={true}
        >
          Enter your iCloud backup password
        </Text>
        <Text
          allowFontScaling={true}
          maxFontSizeMultiplier={1.4}
          style={
            {
              "color": "#7D7D7D",
              "fontFamily": "Basel-Book",
              "fontSize": 17,
              "lineHeight": 24,
              "textAlign": "center",
            }
          }
          suppressHighlighting={true}
        >
          This password is required to recover your recovery phrase backup from iCloud.
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "stretch",
            "flex": 1,
            "flexDirection": "column",
            "flexGrow": 1,
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "stretch",
              "flexDirection": "column",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#F9F9F9",
                "borderBottomColor": "#2222220D",
                "borderBottomLeftRadius": 12,
                "borderBottomRightRadius": 12,
                "borderBottomWidth": 1,
                "borderLeftColor": "#2222220D",
                "borderLeftWidth": 1,
                "borderRightColor": "#2222220D",
                "borderRightWidth": 1,
                "borderStyle": "solid",
                "borderTopColor": "#2222220D",
                "borderTopLeftRadius": 12,
                "borderTopRightRadius": 12,
                "borderTopWidth": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "minHeight": 48,
                }
              }
            >
              <TextInput
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect={false}
                autoFocus={true}
                blurOnSubmit={false}
                clearTextOnFocus={false}
                editable={true}
                focusable={true}
                onChangeText={[Function]}
                onSubmitEditing={[Function]}
                outlineWidth={0}
                placeholder="Enter password"
                placeholderTextColor="#CECECE"
                returnKeyType="done"
                secureTextEntry={true}
                selectionColor="#CECECE"
                style={
                  {
                    "backgroundColor": "transparent",
                    "borderBottomColor": "transparent",
                    "borderBottomLeftRadius": 12,
                    "borderBottomRightRadius": 12,
                    "borderBottomWidth": 0,
                    "borderLeftColor": "transparent",
                    "borderLeftWidth": 0,
                    "borderRightColor": "transparent",
                    "borderRightWidth": 0,
                    "borderStyle": "solid",
                    "borderTopColor": "transparent",
                    "borderTopLeftRadius": 12,
                    "borderTopRightRadius": 12,
                    "borderTopWidth": 0,
                    "color": "#222222",
                    "flex": 1,
                    "fontFamily": "Basel-Book",
                    "fontSize": 17,
                    "height": "auto",
                    "minWidth": 0,
                    "paddingBottom": 12,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 12,
                  }
                }
                textContentType="none"
                value=""
              />
              <View
                style={
                  {
                    "alignItems": "stretch",
                    "flexDirection": "column",
                    "marginLeft": 12,
                    "marginRight": 12,
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  hitSlop={
                    {
                      "bottom": 5,
                      "left": 5,
                      "right": 5,
                      "top": 5,
                    }
                  }
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                      "paddingBottom": 4,
                      "paddingLeft": 4,
                      "paddingRight": 4,
                      "paddingTop": 4,
                    }
                  }
                >
                  <svg
                    color="#7D7D7D"
                    data-file-name="SvgEyeOff"
                    height={20}
                    width={20}
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "stretch",
              "flexDirection": "column",
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#FC72FF",
                "borderBottomColor": "transparent",
                "borderBottomLeftRadius": 16,
                "borderBottomRightRadius": 16,
                "borderBottomWidth": 1,
                "borderLeftColor": "transparent",
                "borderLeftWidth": 1,
                "borderRightColor": "transparent",
                "borderRightWidth": 1,
                "borderStyle": "solid",
                "borderTopColor": "transparent",
                "borderTopLeftRadius": 16,
                "borderTopRightRadius": 16,
                "borderTopWidth": 1,
                "flexDirection": "row",
                "gap": 8,
                "height": "auto",
                "justifyContent": "center",
                "opacity": 0.4,
                "paddingBottom": 12,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 12,
              }
            }
            testID="submit"
            userSelect="none"
          >
            <Text
              maxFontSizeMultiplier={1.2}
              style={
                {
                  "color": "#FFFFFF",
                  "fontFamily": "Basel-Medium",
                  "fontSize": 19,
                  "fontWeight": "500",
                  "lineHeight": 24,
                }
              }
              suppressHighlighting={true}
            >
              Continue
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
